src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScannerTest.java

```diff
@@ -614,20 +778,86 @@ private void testParallelFindsLast(WorkflowJob job, String semaphoreName) throws
         WorkflowRun run  = job.scheduleBuild2(0).getStartCondition().get();
         SemaphoreStep.waitForStart(semaphoreName+"/1", run);
 
-            /*if (run.getExecution() == null) {
-                Thread.sleep(1000);
-            }*/
-
         TestVisitor visitor = new TestVisitor();
-        scan.setup(run.getExecution().getCurrentHeads());
+        List<FlowNode> heads = run.getExecution().getCurrentHeads();
+        scan.setupSorted(heads);
+
+        // Check the sorting order
+        Assert.assertEquals(run.getExecution().getCurrentHeads().size()-1, scan.currentParallelStart.unvisited.size());
+
+        // Check visitor handling
         scan.visitSimpleChunks(visitor, labelFinder);
-        TestVisitor.CallEntry entry = visitor.calls.get(0);
-        Assert.assertEquals(TestVisitor.CallType.CHUNK_END, entry.type);
-        FlowNode lastNode = run.getExecution().getNode(Integer.toString(entry.ids[0]));
-        Assert.assertEquals("Wrong End Node: ("+lastNode.getId()+") "+lastNode.getDisplayName(), "semaphore", lastNode.getDisplayFunctionName());
+        TestVisitor.CallEntry parallelEnd = visitor.calls.get(0);
+        Assert.assertEquals(TestVisitor.CallType.PARALLEL_END, parallelEnd.type);
 
+        TestVisitor.CallEntry entry = visitor.calls.get(1);
+        //FIXME figure out why the chunk_end event doesn't fire, and hack in a way to work around needing to order events here.
```

Comment:
Yes, this is the item in my to-dos.  I'm addressing it along with the review feedback because or a PR this size there will be some.